ngraph/core/src/op/loop.cpp

```diff
@@ -0,0 +1,350 @@
+                // infer type for body_parameter
+                if (body_param_partial_shape.is_dynamic())
+                {
+                    body_parameter->set_partial_shape(input_shape);
+                }
+            }
+        }
+        else if (auto invariant_input_description =
+                     as_type_ptr<TensorIterator::InvariantInputDescription>(input_description))
+        {
+            auto body_parameter =
+                m_body->get_parameters().at(invariant_input_description->m_body_parameter_index);
+
+            auto body_param_partial_shape = body_parameter->get_partial_shape();
+            auto input_partial_shape = input(index).get_partial_shape();
+            NODE_VALIDATION_CHECK(this,
+                                  input_partial_shape.compatible(body_param_partial_shape),
+                                  "Iterator initial value is not compatible with body param");
+
+            if (input_partial_shape.is_static())
+            {
+                auto input_shape = input_partial_shape.to_shape();
+                // infer type for m_body_parameter
+                if (body_param_partial_shape.is_dynamic())
+                {
+                    body_parameter->set_partial_shape(input_shape);
+                }
+            }
+        }
+    }
+
+    // Body
+    m_body->validate_nodes_and_infer_types();
+
+    // Output
+    index_it = 0;
+    for (const auto& output_description : m_output_descriptions)
+    {
+        auto index = output_description->m_output_index;
+        NODE_VALIDATION_CHECK(this, index == index_it, "Output_index not in order");
+        index_it++;
+
+        auto body_value =
+            m_body->get_results().at(output_description->m_body_value_index)->input_value(0);
+
+        if (auto concat_output_description =
+                as_type_ptr<TensorIterator::ConcatOutputDescription>(output_description))
+        {
+            const auto& body_value_partial_shape = body_value.get_partial_shape();
+            set_output_type(index, body_value.get_element_type(), PartialShape::dynamic());
+            if (body_value_partial_shape.is_static())
+            {
+                auto body_value_shape = body_value_partial_shape.to_shape();
+                auto axis = concat_output_description->m_axis;
+
+                Shape out_shape{body_value_shape};
+
+                if (body_value_shape.empty())
+                {
+                    NODE_VALIDATION_CHECK(
+                        this,
+                        axis == 0,
+                        "Axis must be equal to 0 if concatenated output tensor slices are scalars. "
+                        "Loop output index: ",
+                        index);
+                    out_shape = Shape(1);
+                }
+
+                if (m_num_iterations != -1)
+                {
+                    out_shape[axis] = m_num_iterations * body_value_shape[axis];
```

Comment:
`body_value` can be scalar (`body_value_shape == {}`) and it is not handled

Reply:
on the ONNX importer side, we have to insert Unsqueeze op along the concat axis for each such output.
Besides that, there is paranoid handling of this case on lines 229-240.